src/basins/edgetracking.jl

```diff
@@ -0,0 +1,194 @@
+from each other by more than `eps2` (Euclidean distance). The two final states of the
+parallel integration are then used as new states `u1` and `u2` for a new bisection, and 
+so on, until a stopping criterion is fulfilled.
+
+## Output 
+Output can be controlled via the `output_level` argument.
+* `output_level = 0`: returns only the final state on the edge
+* `output_level = 1`: returns the track along the edge as a vector of states
+* `output_level = 2`: returns `[track, bisect1, bisect2]`, where `track` is the track along  
+  the edge and `bisect1`, `bisect2` are vectors containing the bisection points on either  
+  side of the basin boundary
+
+!!! warning
+    May behave erroneously when the DiffEq solver of `ds` is set to `alg = SimpleATsit5()`,
+    which is the default solver in DynamicalSystems. The recommended solver is `Vern9()`.
+
+## References
+[^1]: [Skufca et al. (2006)](https://journals.aps.org/prl/pdf/10.1103/PhysRevLett.96.174101?casa_token=RUn26KnFdNEAAAAA%3AoXsTlmEWVMkEYbOtR-j2PH2vYOOPOy1a2R_37ncnf4gsiHp6GR66M-IBpzXocLoMQC_oHhk8MIFRa_8)
+[^2]: [Schneider et al. (2008)](https://journals.aps.org/pre/pdf/10.1103/PhysRevE.78.037301?casa_token=mLQmTv_cBGUAAAAA%3AVKnQs290sq1MNm-5k8hW7nJeLtVX54I7l-SEGol_HUPSCwziPi-EGDE8ucrDiVMIXZGUbzzam8benFw)
+[^3]: [Luarini and Bodai (2017)](https://iopscience.iop.org/article/10.1088/1361-6544/aa6b11)
+"""
+function edgetracking(ds::DynamicalSystem, u1, u2, attractors::Dict;
+    eps1=1e-7,
+    eps2=1e-6,
+    maxiter=100,
+    abstol=0.0,
+    tmax=Inf,
+    dt=0.01,
+    ϵ_mapper=0.1,
+    output_level=2,
+    verbose=false,
+    kwargs...)
+    
+    pds = ParallelDynamicalSystem(ds, [u1, u2])
+    mapper = AttractorsViaProximity(ds, attractors, ϵ_mapper; kwargs...)
+    
+    edgetracking(pds, mapper;
+        eps1=eps1, eps2=eps2, maxiter=maxiter, abstol=abstol, dt=dt, tmax=tmax,
+        output_level=output_level, verbose=verbose)
+end;
+
+"""
+    edgetracking(pds::ParallelDynamicalSystem, mapper::AttractorMapper; kwargs...)
+Low-level function for the [`edgetracking`](@ref) algorithm.
+"""
+function edgetracking(pds::ParallelDynamicalSystem, mapper::AttractorMapper;
+    eps1=1e-9,
+    eps2=1e-8,
+    maxiter=100,
+    abstol=0.0,
+    dt=0.01,
+    tmax=Inf,
+    output_level=2,
+    verbose=false)
+    
+    verbose && println("=== Starting edge tracking algorithm ===")
+    
+    u1, u2 = bisect_to_edge(pds, mapper; eps=eps1)
+    edgestate = (u1 + u2)/2
+    
+    if output_level>0
+        bisect_pts1 = [u1]
+        bisect_pts2 = [u2]
+        edge = [edgestate]
+    end
+    verbose && println("... Iteration 1: Edge at $(edgestate)")
+
+    correction = abstol + 1
+    counter = 1
+    # edge track iteration loop
+    while (correction > abstol) & (maxiter > counter)
+        reinit!(pds, [u1,u2])
+        state = edgestate
+        distance = dist(current_state(pds, 1) - current_state(pds, 2))
```

Comment:
Alright, this is a comment I'll only leave once here, but it is an absolutely crucial comment that you need to take into account every single time you are doing an operation between the two states. **Your operations are written in suboptimal performance because they would always allocate new vectors if the state of the system is `Vector`**. Said differently, if `x = [1, 2, 3]` then the operation `x + x` is allocating a **new vector and this is unecessary because you don't actually use this vector anywhere, you just take its norm**.

You need to make a function `diffnorm` (for difference norm) that does an explicitly loop, like e.g., we do here:

```julia
function diffnorm(ds::ParallelDynamicalSystem)
    u1 = current_state(ds, 1)
    u2 = current_state(ds, 2)
    # Compute euclidean dinstace in a loop (don't care about static or not)
    d = zero(eltype(u1))
    @inbounds for i in eachindex(u1)
        d += (u1[i] - u2[i])^2
    end
    return sqrt(d)
end
```

Reply:
Good to know! I've created a diffnorm function as suggested.